test/unit/Changelog/ReleaseChangelogViaKeepAChangelogTest.php

```diff
@@ -0,0 +1,212 @@
+        $this->push->expects($this->never())->method('__invoke');
+
+        self::assertNull(
+            $this->releaseChangelog->__invoke(
+                $this->createMockRepositoryWithChangelog(self::INVALID_CHANGELOG),
+                SemVerVersion::fromMilestoneName('1.0.0'),
+                BranchName::fromName('1.0.x')
+            )
+        );
+    }
+
+    public function testNoOpWhenFailedToSetReleaseDateInChangelogEntry(): void
+    {
+        $this->commitFile->expects($this->never())->method('__invoke');
+        $this->push->expects($this->never())->method('__invoke');
+
+        self::assertNull(
+            $this->releaseChangelog->__invoke(
+                $this->createMockRepositoryWithChangelog(self::RELEASED_CHANGELOG),
+                SemVerVersion::fromMilestoneName('1.0.0'),
+                BranchName::fromName('1.0.x')
+            )
+        );
+    }
+
+    public function testWritesCommitsAndPushesChangelogWhenFoundAndReadyToRelease(): void
+    {
+        $existingChangelog = sprintf(self::READY_CHANGELOG, 'TBD');
+        $expectedChangelog = sprintf(self::READY_CHANGELOG, date('Y-m-d'));
+        $repositoryPath    = $this->createMockRepositoryWithChangelog($existingChangelog);
+        $sourceBranch      = BranchName::fromName('1.0.x');
+
+        $this->commitFile
+            ->expects($this->once())
+            ->method('__invoke')
+            ->with(
+                $this->equalTo($repositoryPath),
+                $this->equalTo($sourceBranch),
+                $this->equalTo('CHANGELOG.md'),
+                $this->equalTo('1.0.0 readiness')
+            );
+
+        $this->push
+            ->expects($this->once())
+            ->method('__invoke')
+            ->with(
+                $this->equalTo($repositoryPath),
+                $this->equalTo('1.0.x'),
+            );
+
+        self::assertNull(
+            $this->releaseChangelog->__invoke(
+                $repositoryPath,
+                SemVerVersion::fromMilestoneName('1.0.0'),
+                BranchName::fromName('1.0.x')
+            )
+        );
+
+        $this->assertStringEqualsFile($repositoryPath . '/CHANGELOG.md', $expectedChangelog);
+    }
+
+    /**
+     * @psalm-return non-empty-string
+     */
+    private function createMockRepositoryWithChangelog(
+        string $template,
+        string $filename = 'CHANGELOG.md'
+    ): string {
+        $repo = tempnam(sys_get_temp_dir(), 'ReleaseChangelogViaKeepAChangelog');
+        Assert::notEmpty($repo);
+        unlink($repo);
+
+        (new Process(['mkdir', '-p', $repo]))->mustRun();
+
+        file_put_contents(
+            sprintf('%s/%s', $repo, $filename),
+            $template
+        );
+
+        return $repo;
+    }
+
+    private const INVALID_CHANGELOG = <<< 'END'
+        # NOT A CHANGELOG
+
+        This file is not a changelog.
+
+        ## Bad headers
+
+        It contains bad headers, among other things.
+
+        END;
```

Comment:
Since we're on 7.4... One cool feature of 7.3 is that it will now strip indentation from heredoc/nowdoc contents if the end marker is indented. Makes these constant declarations far more readable!

Reply:
I would still endorse using a different start/end marker depending on context. For example, here I would have used `<<< 'MARKDOWN'`

Not important, since we're in the middle of a relatively clear test